packages/snaps-controllers/src/services/webview/WebViewMessageStream.ts

```diff
@@ -0,0 +1,101 @@
+import type { PostMessageEvent } from '@metamask/post-message-stream';
+import { BasePostMessageStream } from '@metamask/post-message-stream';
+import { isValidStreamMessage } from '@metamask/post-message-stream/dist/utils';
+import { logError } from '@metamask/snaps-utils';
+import { assert } from '@metamask/utils';
+
+// TBD
+export type WebViewInterface = {
+  injectJavaScript(js: string): void;
+  addEventListener(
+    type: 'message',
+    listener: (event: MessageEvent) => void,
+  ): void;
+  removeEventListener(
+    type: 'message',
+    listener: (event: MessageEvent) => void,
+  ): void;
+};
+
+type WebViewStreamArgs = {
+  name: string;
+  target: string;
+  getWebView: () => Promise<WebViewInterface>;
+};
+
+/**
+ * A special postMessage stream used to interface with a WebView.
+ */
+
+export class WebViewMessageStream extends BasePostMessageStream {
+  #name;
+
+  #target;
+
+  #webView: WebViewInterface | undefined;
+
+  /**
+   * Creates a stream for communicating with other streams inside a WebView.
+   *
+   * @param args - Options bag.
+   * @param args.name - The name of the stream. Used to differentiate between
+   * multiple streams sharing the same window object.
+   * @param args.target - The name of the stream to exchange messages with.
+   * @param args.getWebView - A asynchronous getter for the webview.
+   */
+  constructor({ name, target, getWebView }: WebViewStreamArgs) {
+    super();
+
+    this.#name = name;
+    this.#target = target;
+
+    this._onMessage = this._onMessage.bind(this);
+
+    // This is a bit atypical from other post-message streams.
+    // We have to wait for the WebView to fully load before we can continue using the stream.
+    getWebView()
+      .then((webView) => {
+        this.#webView = webView;
+        webView.addEventListener(
+          'message',
+          (event: MessageEvent<PostMessageEvent>) =>
+            this._onMessage(event as any),
+        );
+        this._handshake();
+      })
+      .catch((error) => {
+        logError(error);
+      });
+  }
+
+  protected _postMessage(data: unknown): void {
+    assert(this.#webView);
+    this.#webView.injectJavaScript(
+      `window.postMessage(${JSON.stringify({
+        target: this.#target,
+        data,
```

Comment:
(high severity) XSS sink.
This is the only severe risk we're taking here. 
`data` may be arbitrary data coming from any dapp. If it contains a sequence that can be misinterpreted by the engine within webview to end the current string content, we get arbitrary code execution within the webview. 

It does seem highly unlikely, but remember we're looking at stringify implementation of one engine creating a string that gets executed in a different engine and the differences in how the weirder utf-8 characters can be treated existed historically.

To avoid the need to fuzz the implementation, we'd be better off encoding the message data in a format that does not contain bytes from the original string directly, eg. by putting the stringified data in a buffer and encoding it to a base64 string.
I'd recommend researching which built-in capability in the webview would allow for the fastest decoding of whatever encoding we want to use. Buffer polyfill is unlikely to be the fastest option.

Performance is a factor here, so transfering the source code of the snap could be an exception to handle performance differently (with a limited set of characters allowed maybe?)


Side note: 
It might be possible to thread the communication through fetch or similar, with the mobile app "serving" content. That would avoid injecting javascript and having a global sink for messages altogether.

Reply:
@naugtur I just pushed a commit that uses base64 for encoding the messages.

@Jonathansoufer We should do performance testing to make sure this still works reasonably well on mobile devices. We can consider using native code for the encoding if necessary.